tokio/src/net/tcp/stream.rs

```diff
@@ -936,6 +936,168 @@ impl TcpStream {
             .try_io(Interest::WRITABLE, || (&*self.io).write_vectored(bufs))
     }
 
+    /// Try to read from the socket using a user-provided IO operation.
+    ///
+    /// If the socket is ready for reading, the provided closure is called. The
+    /// closure should attempt to read from the socket by manually calling the
+    /// appropriate syscall. If the operation fails because the socket is not
+    /// actually ready, then the closure should return a `WouldBlock` error and
+    /// the read readiness flag is cleared. The return value of the closure is
+    /// then returned by `try_read_io`.
+    ///
+    /// If the socket is not ready for reading, then the closure is not called
+    /// and a `WouldBlock` error is returned.
+    ///
+    /// The closure should only return a `WouldBlock` error if it has performed
+    /// an IO operation on the socket that failed due to the socket not being
+    /// ready. Returning a `WouldBlock` error in any other situation will
+    /// incorrectly clear the readiness flag, which can cause the socket to
+    /// behave incorrectly.
+    ///
+    /// The closure should not perform the read operation using any of the
+    /// methods defined on the Tokio `TcpStream` type, as this will mess with
+    /// the readiness flag and can cause the socket to behave incorrectly.
+    ///
+    /// Usually, [`readable()`] or [`ready()`] is used with this function.
+    ///
+    /// [`readable()`]: TcpStream::readable()
+    /// [`ready()`]: TcpStream::ready()
+    pub fn try_read_io<R>(&self, f: impl FnOnce() -> io::Result<R>) -> io::Result<R> {
+        self.io.registration().try_io(Interest::READABLE, f)
+    }
+
+    /// Polls for read from the socket using a user-provided IO operation.
```

Comment:
The user-provided IO operation does not perform the polling.
```suggestion
    /// Poll for read readiness, then reads from the socket using a user-provided IO operation.
```